clang/lib/3C/RewriteUtils.cpp

```diff
@@ -229,11 +229,52 @@ static void emit(Rewriter &R, ASTContext &C) {
       };
 
       // Check whether we are allowed to write this file.
+      //
+      // In our testing as of 2021-03-15, the file path returned by
+      // FE->tryGetRealPathName() was canonical, but be safe against unusual
+      // situations or possible future changes to Clang before we actually write
+      // a file. We can't use FE->getName() because it seems it may be relative
+      // to the `directory` field of the compilation database, which (now that
+      // we no longer use `ClangTool::run`) is not guaranteed to match 3C's
+      // working directory.
       std::string ToConv = FE->tryGetRealPathName().str();
       std::string FeAbsS = "";
-      getCanonicalFilePath(ToConv, FeAbsS);
+      std::error_code EC = tryGetCanonicalFilePath(ToConv, FeAbsS);
+      if (EC) {
+        unsigned ErrorId = DE.getCustomDiagID(
+            UnwritableChangeDiagnosticLevel,
+            "3C internal error: not writing the new version of this file due "
+            "to failure to re-canonicalize the file path provided by Clang");
+        DE.Report(SM.translateFileLineCol(FE, 1, 1), ErrorId);
+        {
```

Comment:
You shouldn't need this code block. As invaluable as this was when we were calling `exit()`, diagnostics will be emitted properly when exiting naturally, which is done in 3C.cpp through tool.run() or when diagnostics error count exceeds 0.
And it's within an `if` block, which takes care of it.

And it looks like I forgot to remove the block in ProgramInfo.cpp myself after adding the comment below it for why the `exit()` is unnecessary.

Reply:
Without the block, this `DiagnosticBuilder` will still exist during the call to `PrintExtraUnwritableChangeInfo` and Clang raises an assertion failure about two `DiagnosticBuilder`s existing at the same time.  There is another case like this in `rewriteSourceRange`.  Should we comment every such case?

Reply:
There's prior code that didn't have this trouble. Is it about the error together with the note? How does clang (or other) deal with this situation? Does the call to `getDiagnostics()` close open builders?

Reply:
> There's prior code that didn't have this trouble.

Can you point me to this code?  That might be the fastest way to clear this up.

> Is it about the error together with the note?

No, for the error, the `DiagnosticBuilder` returned by `DE.Report` isn't saved to a variable, so it gets destructed immediately.  The problem is that the `DiagnosticBuilder` for the note is still held in a local variable while additional diagnostics are generated by `PrintExtraUnwritableChangeInfo`.

Reply:
Sorry, I missed the importance of the local variable for both the old and new code. You could consider moving the call to `PrintExtraUnwritableChangeInfo` before the builder, but then I'm out of ideas and this is fine as it is. Yes, please add a brief comment on the block if you keep it.

Reply:
> You could consider moving the call to `PrintExtraUnwritableChangeInfo` before the builder

I don't want to do that because the note provides values specifically related to this error (I thought it would be too long to put everything in one error diagnostic) while the `PrintExtraUnwritableChangeInfo` boilerplate is the same for 4 different error cases.

> Yes, please add a brief comment on the block if you keep it.

Done.  Maybe in the future, [we can use a more concise wrapper function to generate diagnostics](https://github.com/correctcomputation/checkedc-clang/pull/488/#discussion_r599546827) and we won't have to worry about these blocks.